### ISSUE TITLE
Resumability "handshake"

### ISSUE THREAD

**Issue by NiteshKant** [ReactiveSocket/reactivesocket#144]
Carrying over the [discussion from the PR](https://github.com/ReactiveSocket/reactivesocket/pull/135#discussion_r94722291) here.

#### Problem
When a server does not support resumability, the protocol mandates that it should return a [setup error](https://github.com/ReactiveSocket/reactivesocket/blob/master/Protocol.md#setup-frame). Such a behavior creates this awkward interaction for clients that can support resumable and non-resumable streams:

- Connect to the server with SETUP indicating resumability required.
- Server rejects setup which leads to shutdown of the connection.
- Client reconnects with resumability disabled.

The above assumes that resumability is optional for the client (as suggested by the protocol)

It appears that the following options are available for this interaction:

- Explicit ack for setup (suggesting resume supported/unsupported and potentially some other optional setup properties)
- Reject setup if resume not supported by a server.
- "Soft error" for resume not supported by a server and let the client decide whether to continue or disconnect.

IMO, third option above is better but we have chosen second. What are the reasons for choosing the second option and are there any drawbacks in choosing the third?

**Comment by NiteshKant** [comment]
\cc @benjchristensen @tmontgomery

**Comment by benjchristensen** [comment]
I see the point if resumability is optional for a use case. We don't have a concept of "soft error" in the protocol right now however, so how do you suggest making that happen?

However, I don't see this being a big deal in general, as a client expecting resumability is not going to be connecting in prod to servers that randomly reject resumability in the SETUP. They either support it or don't. For that reason, I'm okay leaving it as it is now. To me it is an exceptional case found in development, not production, trying to talk to a server that doesn't support resumability. 

Where there can be random behavior is on the RESUME, where servers may or may not be able to resume, and that flow already supports an ACK/ERROR. 

In other words, resumability being rejected at SETUP is a statement that the server does not support it it all, not that it may not be able to resume later, which it can't generally know ahead of time. And a server isn't going to randomly change that in production.

**Comment by yschimke** [comment]
I think some of this depends on how generic we think these libraries are.  Whether server APIs will be deployed, proxies/load balancers used or generic client libraries used on mobile clients. HTTP faces many of the same issues but has to take an extreme position of negotiation, upgrades and tolerance of version ranges.

Another example is a new client (that supports multiple versions) connecting to an old server.  presumably it needs to walk down the version ladder until it finds a match.

**Comment by tmontgomery** [comment]
One way to get an ACK is to use a LEASE to get a response. And just set the value really high.

**Comment by benjchristensen** [comment]
Since the transport isn't torn down on an error, sending a CONNECTION_ERROR and then the client re-sending another SETUP seems to be no more or less expensive than an ACK/ERROR approach. 

So why not just let CONNECTION_ERROR, SETUP, and RESUME behave as they are now?

**Comment by NiteshKant** [comment]
Here is from the protocol:

```
CONNECTION_ERROR	0x00000101	The connection is being terminated. Stream ID MUST be 0.
```

So, if the server sends a `CONNECTION_ERROR` then the connection is practically closed.

This is how the interaction would look like for "soft-error":

- Client sends SETUP with resume enabled.
- Server sends a `RESUME_NOT_SUPPORTED` error on Stream 0. Protocol defines this as an error which isn't fatal and hence does not close connection.
- Client decides whether to continue (degraded and not expecting resume) or not.
- If decides to abandon, closes the connection (`CONNECTION_ERROR` on stream 0)

**Comment by benjchristensen** [comment]
> "the connection"

The ReactiveSocket connection, sure, but that doesn't mean the transport "connection".

**Comment by benjchristensen** [comment]
> Protocol defines this as an error which isn't fatal and hence does not close connection.

For me this just complicates the implementation of the protocol. It means it now needs state about errors and continuing. Instead just die, restart. The over the wire cost is the same. It's a new frame being sent. 

Receiving an error frame on a ReactiveSocket does not mean the underlying transport is shut down and re-established. In fact the implementations we're using right now absolutely do not tear own the transports. They reuse them for many ReactiveSocket connections.

**Comment by NiteshKant** [comment]
>>> They reuse them for many ReactiveSocket connections.

Interesting and the stream IDs are non-overlapping for these RS connections?

**Comment by yschimke** [comment]
Both of these are true

- With resumable ReactiveSocket, a socket can have in its lifetime 0, 1 or many transport connections (DuplexConnection).  At any time the socket has at most one DuplexConnection.
- A DuplexConnection may use a single stream on a multiplexed network connection e.g. HTTP/2 connection that carries other ReactiveSockets or regular HTTP traffic.

But we don't have cases were a single DuplexConnection is reused across multiple ReactiveSocket instances.

**Comment by benjchristensen** [comment]
> Interesting and the stream IDs are non-overlapping for these RS connections?

Concurrent ReactiveSocket stream IDs can be overlapping when we are on an HTTP/2 transport, as the multiplexing occurs on top of HTTP/2 streams.

**Comment by NiteshKant** [comment]
If I am the only one who feels that this is an issue, we can close this issue 😄

**Comment by benjchristensen** [comment]
Anyone else think this is an issue?

**Comment by tmontgomery** [comment]
I do not think this is an issue.

**Comment by benjchristensen** [comment]
Closing.
